restli-client/src/test/java/com/linkedin/restli/client/TestParSeqBasedCompletionStage.java

```diff
@@ -0,0 +1,1363 @@
+
+    BiFunction<String, Throwable, Integer> consumer = (s, throwable) -> {
+      throw EXCEPTION;
+    };
+
+    Function<Throwable, Integer> errorHandler = mock(Function.class);
+    finish(completionStage.handle(consumer).exceptionally(errorHandler));
+
+    verify(errorHandler).apply(verifyCompletionException());
+  }
+
+  @Test public void testExceptionPropagation_shouldNotFailOnThrowable() throws Exception{
+    Consumer<Integer> intConsumer = mock(Consumer.class);
+    CompletionStage<String> completionStage = createTestFailedStage(EXCEPTION);
+
+    Function<Throwable, Void> errorFunction = mock(Function.class);
+    finish(completionStage.thenApply(String::length).thenApply(i -> i * 2).thenAccept(intConsumer).exceptionally(errorFunction));
+
+
+    verifyZeroInteractions(intConsumer);
+    verify(errorFunction, times(1)).apply(verifyCompletionException());
+
+
+  }
+
+  @Test
+  public void testWhenComplete() throws Exception
+  {
+    BiConsumer<String, Throwable> biConsumer = mock(BiConsumer.class);
+    CompletionStage<String> stage = createTestStage(TESTVALUE1).whenComplete(biConsumer);
+    finish(stage);
+    verify(biConsumer).accept(TESTVALUE1, null);
+  }
+
+  @Test
+  public void testWhenComplete_withException() throws Exception
+  {
+    BiConsumer<String, Throwable> biConsumer = mock(BiConsumer.class);
+    CompletionStage<String> stage = createTestFailedStage(EXCEPTION).whenComplete(biConsumer);
+    finish(stage);
+    verify(biConsumer, times(1)).accept(null, EXCEPTION);
+  }
+
+  @Test
+  public void testWhenComplete_useUnwrappedException() throws Exception
+  {
+    BiConsumer<String, Throwable> biConsumer = mock(BiConsumer.class);
+    CompletionStage<String> completionStage = createTestFailedStage(EXCEPTION);
+    finish(completionStage.whenComplete(biConsumer));
+    verify(biConsumer, times(1)).accept(null, EXCEPTION);
+  }
+
+  @Test
+  public void testWhenComplete_completeWithException() throws Exception
+  {
+    BiConsumer<String, Throwable> consumer = (v, t) -> {
+      throw EXCEPTION;
+    };
+
+    Function<Throwable, String> exceptionallyFunction = mock(Function.class);
+
+    CompletionStage<String> completionStage = createTestStage(TESTVALUE1);
+    finish(completionStage.whenComplete(consumer).exceptionally(exceptionallyFunction));
+    verify(exceptionallyFunction, times(1)).apply(verifyCompletionException());
+  }
+
+  @Test
+  public void testWhenComplete_handleExceptionWithSuccess() throws Exception
+  {
+    // Also test propagating
+
+    Function<Throwable, String> exceptionallyFunction = mock(Function.class);
+
+    CompletionStage<String> completionStage = createTestFailedStage(EXCEPTION);
+    finish(completionStage.whenComplete((v, t) -> {}).exceptionally(exceptionallyFunction));
+    verify(exceptionallyFunction, times(1)).apply(verifyCompletionException());
+  }
+
+  @Test
+  public void testWhenCompleteAsync() throws Exception
+  {
+    CountDownLatch waitLatch = new CountDownLatch(1);
```

Comment:
You don't need to use a latch. Just convert to completable future and use the get with timeout.

Reply:
Yes I used this latch extensively in this file and I know some of them are not needed, or to be replaced by your suggestion.

But others are necessary, they are not just for waiting the result: I want to make sure the latch has been counted down, which could prove that the logic actually executed. (You know sometimes the completion stage can completes with failure even without the whole logic executed. ) 

I can do similar thing with checking a counter? But I think latch makes more sense in these tests